consensus/src/payload_generator.rs

```diff
@@ -0,0 +1,367 @@
+            if prev_sender != Some(current_sender) {
+                let txn_chunk = TransactionsChunkBySender::new(current_sender);
+                candidate_txn_chunks.push_back(txn_chunk);
+            }
+            candidate_txn_chunks
+                .back_mut()
+                .unwrap()
+                .add_transaction(txn);
+            prev_sender = Some(current_sender)
+        }
+        for i in 0..num_transactions {
+            let remaining_txns = num_transactions - i;
+            let max_lookup = min(self.num_transactions_to_look_ahead, remaining_txns);
+            let mut to_be_pushed_back_chunks = VecDeque::new();
+            let mut candidate_found = false;
+            let mut current_index = 0;
+            while current_index < max_lookup {
+                let mut candidate_chunk = candidate_txn_chunks
+                    .pop_front()
+                    .expect("Expected transaction chunk in the candidate transaction chunk");
+                if !self
+                    .sliding_window
+                    .has_conflict_in_window(&candidate_chunk.sender())
+                {
+                    self.sliding_window.add_transaction(
+                        candidate_chunk
+                            .remove_transaction()
+                            .expect("Expected transaction in candidate chunk"),
+                    );
+                    candidate_found = true;
+                    if !candidate_chunk.is_empty() {
+                        to_be_pushed_back_chunks.push_front(candidate_chunk)
+                    }
+                    break;
+                }
+                current_index += candidate_chunk.len();
+                to_be_pushed_back_chunks.push_front(candidate_chunk);
+            }
+
+            if !candidate_found {
+                // We didn't find any non-conflicting txn in the look up window. In this case, just
+                // add the first candidate to the block.
+                let mut chunk = to_be_pushed_back_chunks
+                    .pop_back()
+                    .expect("Expected non empty vector");
+                self.sliding_window.add_transaction(
+                    chunk
+                        .remove_transaction()
+                        .expect("Empty chunk not expected"),
+                );
+                if !chunk.is_empty() {
+                    to_be_pushed_back_chunks.push_back(chunk)
+                }
+            }
+
+            // Add the remaining txns to the candidate txns list in the original order.
+            let mut chunk = to_be_pushed_back_chunks.pop_front();
+            while chunk.is_some() {
+                candidate_txn_chunks.push_front(chunk.unwrap());
+                chunk = to_be_pushed_back_chunks.pop_front();
+            }
+        }
+        self.sliding_window.finalize()
+    }
+}
+
+impl SenderAwarePayloadGenerator {
+    pub fn new(conflict_window_size: usize, num_transactions_to_look_ahead: usize) -> Self {
```

Comment:
what is the difference between conflict_window_size and num_transactions_to_look_ahead?

Reply:
conflict_window_size is the size of the past window to maintain when considering for conflict. For example, if in blockSTM, parallelism is 32, then a conflict window can be ~32. 

look_ahead_window is just an optimization to reduce the cost of the look up - i.e., how many transactions we look ahead in the block before giving up and adding the first transaction even if there is a conflict. 

I will add some documentation around the behavior of the shuffler soon, which should make it clearer hopefully.